README.md
<br />
<h1 align="center">Joy Dropdowns</h1>

<p align="center">A beautiful package to create controls with properties equivalent to web Dropdown quickly and customizable. <a href="https://lincolnjota.github.io/dropdowns-example/#/">see a live preview</a></p>

# Installing

### 1. Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  joy_dropdowns: ^1.0.0+1
```

### 2. Install it

You can install packages from the command line:

with `pub`:

```
$ pub get
```

with `Flutter`:

```
$ flutter pub get
```

### 3. Import it

Now in your `Dart` code, you can use:

```dart
import  'package:joy_dropdowns/joy_dropdowns.dart';
```

# Usage
```dart
JoyDropDown(
  overlayHeight:  300,
  overlayWidth:  250,
  overlayOffset:  const  Offset(0, 5),
  overlayPosition:  JoyOverlayPosition.bottom,
  overlayWidget:  Container(
    decoration:  const  BoxDecoration(
      color:  Colors.red,
	),
    child:  SingleChildScrollView(
      physics:  const  ScrollPhysics(),
	  child:  ListView.builder(
          shrinkWrap:  true,
          itemCount:  4,
          itemBuilder: (context, index) {
		    return  InkWell(
		      onTap: () {
		        print('$index');
			  },
              child:  ListTile(
                leading:  const  Icon(Icons.ac_unit),
                title:  Text('Item $index'),
              ),
          );
        },
      ),
    ),
  ),
  child:  ElevatedButton(
    onPressed: () {},
    child:  Text('hover-me'),
   ),
),
```

These are all settings, but it goes from imagination to creating the other overlays.

- `overlayHeight` – This setting is self explanatory, but if it is null, the height will be set by the child of the button.
- `overlayWidth` – Same intention as the previous one. If it is null, the width will be set by the child of the button.
- `overlayOffset` – Maybe you need to move a little bit over there, a little bit over here...
- `overlayPosition` –Where do you have the dropdown appear? That enum can help you. possible are: bottom, left, top, right.
- `overlayWidget` – This is the dropdown widget, you need to put something other than a list, I put open for creativity.
- - `child` – This is the hoverable button widget, it can be anything, a custom container, an ElevatedButton...


## JoyMenu - Fast & Elegant list

If you want to add only one list and be happy,  there is a factory method 
called "JoyDropDown.list", see:

<img src="https://i.imgur.com/U66W1fQ.png" align = "right" height = "300px">

```dart
JoyDropDown.list(
  overlayOffset:  const  Offset(0, 5),
  overlayPosition:  JoyOverlayPosition.bottom,
  menu:  JoyMenu(
    menuDecoration:  BoxDecoration(
      boxShadow: [
        BoxShadow(
          color:  Colors.black.withOpacity(0.5),
          blurRadius:  10,
          spreadRadius:  2,
        ),
      ],
      borderRadius:  BorderRadius.circular(10),
    ),
	hoverColor:  Color.fromARGB(255, 0, 116, 104),
	defaultColor:  Colors.teal,
	items:  items,
   ),
   child: ElevatedButton(
     onPressed: () {},
     child:  Text('hover-me'),
   ),
```

**Note:** It has many settings that can be changed. 
### If you need help, make compliments, donations etc. I'll be online at Discord: JoyGDR#6722.


CHANGELOG.md
## [1.0.0+1] Some fixes and improvements.

* Some fixes and improved documentation.

## [1.0.0] Yuupi!

* First release

test/joy_dropdown_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:joy_dropdown/joy_dropdown.dart';

// void main() {
//   // test('adds one to input values', () {
//   //   final calculator = Calculator();
//   //   expect(calculator.addOne(2), 3);
//   //   expect(calculator.addOne(-7), -6);
//   //   expect(calculator.addOne(0), 1);
//   // });
// }


lib/src/button.dart
import 'package:flutter/material.dart';
import 'joy_list.dart';
import 'close_menu.dart';
import 'overlay.dart';
import 'enums.dart';

class JoyDropDown extends StatefulWidget {
  final Widget child;
  final Widget overlayWidget;
  final JoyOverlayPosition? overlayPosition;
  final Offset? overlayOffset;
  final double? overlayWidth;
  final double? overlayHeight;
  final UniqueKey uniqueKey = UniqueKey();

  /// ## JoyDropDown
  /// Creates a widget that has your child as a "hoverable" Here are some properties available:
  /// - ``overlayWidget``: Widget you want to appear at the time of hovering over the mouse..
  /// - ``overlayPosition``: Position of the overlayWidget. Default is JoyOverlayPosition.bottom.
  /// - ``overlayOffset``: Offset of the overlayWidget. Default is Offset(0, 0).
  /// - ``overlayWidth``: If this parameter is not set, it will be the same as the child width.
  /// - ``overlayHeight``: If this parameter is not set, it will be the same as the child height.
  /// - ``child``: Widget that will be hovered. better known as button.
  // ignore: use_key_in_widget_constructors
  JoyDropDown({
    required this.child,
    required this.overlayWidget,
    this.overlayPosition,
    this.overlayOffset,
    this.overlayWidth,
    this.overlayHeight,
  });

  /// ## JoyDropDown.list
  /// Creates a widget that has your child as a "hoverable" Here are some properties available:
  /// - ``menu``: Widget you want to appear at the time of hovering over the mouse..
  /// - ``overlayPosition``: Position of the overlayWidget. Default is JoyOverlayPosition.bottom.
  /// - ``overlayOffset``: Offset of the overlayWidget. Default is Offset(0, 0).
  /// - ``overlayWidth``: If this parameter is not set, it will be the same as the child width.
  /// - ``overlayHeight``: If this parameter is not set, it will be the same as the child height.
  /// - ``child``: Widget that will be hovered. better known as button.
  factory JoyDropDown.list({
    required JoyOverlayPosition overlayPosition,
    required JoyMenu menu,
    required Widget child,
    Offset? overlayOffset,
    double? overlayWidth,
    double? overlayHeight,
  }) {
    return JoyDropDown(
      overlayWidget: menu,
      overlayPosition: overlayPosition,
      overlayOffset: overlayOffset ?? Offset.zero,
      overlayWidth: overlayWidth ?? 200,
      overlayHeight: overlayHeight ?? 200,
      child: child,
    );
  }

  @override
  State<JoyDropDown> createState() => _JoyDropDownState();
}

class _JoyDropDownState extends State<JoyDropDown> {
  LayerLink link = LayerLink();

  @override
  Widget build(BuildContext context) {
    return MouseRegion(
      onEnter: (event) {
        if (closeTimer != null) {
          closeTimer!.cancel();
        }

        if (!isMenuShowing || widget.uniqueKey != currentChildKey) {
          if (overlayEntry != null) {
            isMenuShowing = false;
            overlayEntry!.remove();
            overlayEntry = null;
          }

          showJoyDropDown(
            context: context,
            overlayWidget: widget.overlayWidget,
            overlayPosition:
                widget.overlayPosition ?? JoyOverlayPosition.bottom,
            overlayOffset: widget.overlayOffset ?? Offset.zero,
            overlayHeight: widget.overlayHeight,
            overlayWidth: widget.overlayWidth,
            layerLink: link,
            key: widget.uniqueKey,
          );
        }
      },
      onExit: (event) async {
        closeJoyDropDown(context);
      },
      child: CompositedTransformTarget(
        link: link,
        child: widget.child,
      ),
    );
  }
}


lib/src/overlay.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'close_menu.dart';
import 'enums.dart';

OverlayEntry? overlayEntry;
bool isMenuShowing = false;
Timer? closeTimer;
Key? currentChildKey;

void showJoyDropDown({
  required BuildContext context,
  required Widget overlayWidget,
  required JoyOverlayPosition overlayPosition,
  required LayerLink layerLink,
  required Offset overlayOffset,
  required Key key,
  double? overlayHeight,
  double? overlayWidth,
}) {
  RenderBox? renderObject;

  overlayEntry = OverlayEntry(
    builder: (_) {
      renderObject = context.findRenderObject() as RenderBox;
      Alignment followerAnchor = Alignment.topLeft;
      Alignment targetAnchor = Alignment.topLeft;

      switch (overlayPosition) {
        case JoyOverlayPosition.top:
          followerAnchor = Alignment.bottomCenter; // Ancora do vermelho
          targetAnchor = Alignment.topCenter; // Ancora do azul
          break;
        case JoyOverlayPosition.bottom:
          followerAnchor = Alignment.topCenter; // Ancora do vermelho
          targetAnchor = Alignment.bottomCenter; // Ancora do azul
          break;
        case JoyOverlayPosition.left:
          followerAnchor = Alignment.topRight; // Ancora do vermelho
          targetAnchor = Alignment.topLeft; // Ancora do azul
          break;
        case JoyOverlayPosition.right:
          followerAnchor = Alignment.topLeft; // Ancora do vermelho
          targetAnchor = Alignment.topRight; // Ancora do azul
          break;
        default:
      }

      return Positioned(
        width: overlayWidth ?? renderObject!.size.width,
        height: overlayHeight ?? renderObject!.size.height,
        //left: pos.dx,
        child: CompositedTransformFollower(
          link: layerLink,
          showWhenUnlinked: false,
          offset: overlayOffset,
          followerAnchor: followerAnchor,
          targetAnchor: targetAnchor,
          child: MouseRegion(
            onEnter: (event) {
              if (closeTimer != null) {
                closeTimer!.cancel();
              }
            },
            onExit: (event) {
              closeJoyDropDown(context);
            },
            child: Material(
                borderOnForeground: false,
                type: MaterialType.transparency,
                child: overlayWidget),
          ),
        ),
      );
    },
  );

  Overlay.of(context)!.insert(overlayEntry!);
  currentChildKey = key;
  isMenuShowing = true;
}


lib/src/joy_list.dart
import 'package:flutter/material.dart';
import 'close_menu.dart';

class JoyMenu extends StatefulWidget {
  final List<JoyMenuItem> items;
  final Color? hoverColor;
  final Color? defaultColor;
  final BoxDecoration? menuDecoration;

  /// It is a meth to make beautiful lists, and fast, without the need to configure many decorations.
  /// Widget that can be set in the ``menu`` property of the
  /// ```dart
  /// JoyDropDown.list(menu: JoyMenu(...), ...)
  /// ```
  /// Here are some possible settings:
  /// * ``menuDecoration``: The decoration of the menu. (Its a Container)
  /// * ``hoverColor``: The color of the menu when the mouse is over it.
  /// * ``defaultColor``: The color of the menu when the mouse is not over it.
  /// * ``items``: A list that should be of type List<JoyMenuItem>...
  const JoyMenu({
    Key? key,
    required this.items,
    this.hoverColor,
    this.defaultColor,
    this.menuDecoration,
  }) : super(key: key);

  @override
  State<JoyMenu> createState() => _JoyMenuState();
}

class _JoyMenuState extends State<JoyMenu> {
  int? selectedIndex;

  @override
  Widget build(BuildContext context) {
    return Container(
      clipBehavior: Clip.hardEdge,
      decoration: widget.menuDecoration,
      child: SingleChildScrollView(
        physics: const ScrollPhysics(),
        child: ListView.builder(
          shrinkWrap: true,
          itemCount: widget.items.length,
          itemBuilder: (context, index) {
            return Container(
              //padding: const EdgeInsets.only(left: 4),
              height: 50,
              width: double.infinity,
              decoration: BoxDecoration(
                  color: selectedIndex != index || selectedIndex == null
                      ? widget.defaultColor ??
                          Theme.of(context).colorScheme.primaryContainer
                      : widget.hoverColor ??
                          Theme.of(context).colorScheme.onPrimaryContainer),
              child: Material(
                type: MaterialType.transparency,
                child: InkWell(
                  onTap: () {
                    widget.items[index].onTap();
                    closeJoyDropDown(context);
                  },
                  onHover: (value) {
                    setState(() {
                      selectedIndex = index;
                    });
                  },
                  child: Row(
                    children: [
                      const SizedBox(width: 10),
                      Icon(widget.items[index].icon.icon,
                          color: widget.items[index].icon.color ??
                              Theme.of(context).colorScheme.onBackground,
                          size: widget.items[index].icon.size ?? 20),
                      const SizedBox(width: 10),
                      Text(
                        widget.items[index].title.data ?? '',
                        style: widget.items[index].title.style ??
                            Theme.of(context).textTheme.bodyText2?.copyWith(
                                  color: Theme.of(context)
                                      .colorScheme
                                      .onBackground,
                                ),
                      ),
                    ],
                  ),
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}

class JoyMenuItem {
  /// It's like listile, only it has the properties a navMenu needs!
  /// * ``icon``: The icon of the menu item.
  /// * ``title``: The title of the menu item.
  /// * ``onTap``: The function that will be called when the menu item is tapped.
  JoyMenuItem({
    required this.icon,
    required this.title,
    required this.onTap,
  });

  final Icon icon;
  final Text title;
  final void Function() onTap;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:joy_dropdowns/joy_dropdowns.dart';

import 'src/models/example_button.dart';

void main() {
  runApp(const MaterialApp(
    home: HomePage(),
  ));
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    List<JoyMenuItem> items = [
      JoyMenuItem(
          icon: const Icon(Icons.abc),
          title: const Text('Category'),
          onTap: () => print('Category')),
      JoyMenuItem(
          icon: const Icon(Icons.access_time_filled),
          title: const Text('About me'),
          onTap: () => print('About me')),
      JoyMenuItem(
          icon: const Icon(Icons.abc),
          title: const Text('Buy now'),
          onTap: () => print('Buy now')),
      JoyMenuItem(
          icon: const Icon(Icons.abc),
          title: const Text('anything'),
          onTap: () => print('anything')),
    ];
    return Scaffold(
      appBar: AppBar(
        title: const Text('Dropdowns Examples'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            const Text(
              'Positions',
              style: TextStyle(fontSize: 50),
            ),
            const SizedBox(height: 20),
            Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                // LEFT POSITION
                JoyDropDown(
                  overlayHeight: 300,
                  overlayWidth: 250,
                  overlayOffset: const Offset(-5, 0),
                  overlayPosition: JoyOverlayPosition.left,
                  overlayWidget: Container(
                    decoration: const BoxDecoration(
                      color: Colors.red,
                    ),
                    child: SingleChildScrollView(
                      physics: const ScrollPhysics(),
                      child: ListView.builder(
                        shrinkWrap: true,
                        itemCount: 18,
                        itemBuilder: (context, index) {
                          return InkWell(
                            onTap: () {
                              print('$index');
                            },
                            child: ListTile(
                              leading: const Icon(Icons.ac_unit),
                              title: Text('Item $index'),
                            ),
                          );
                        },
                      ),
                    ),
                  ),
                  child: const ExampleButton(
                      text: 'Left Position', color: Colors.red),
                ),
                const SizedBox(width: 20),
                JoyDropDown(
                  overlayHeight: 300,
                  overlayWidth: 250,
                  overlayOffset: const Offset(5, 0),
                  overlayPosition: JoyOverlayPosition.right,
                  overlayWidget: Container(
                    decoration: const BoxDecoration(
                      color: Colors.purple,
                    ),
                    child: SingleChildScrollView(
                      physics: const ScrollPhysics(),
                      child: ListView.builder(
                          shrinkWrap: true,
                          itemCount: 18,
                          itemBuilder: (context, index) {
                            return InkWell(
                              onTap: () {
                                print('$index');
                              },
                              child: ListTile(
                                leading: const Icon(Icons.ac_unit),
                                title: Text('Item $index'),
                              ),
                            );
                          }),
                    ),
                  ),
                  child: const ExampleButton(
                      text: 'Right Position', color: Colors.purple),
                ),
                const SizedBox(width: 20),
                JoyDropDown.list(
                  overlayOffset: const Offset(0, -5),
                  menu: JoyMenu(
                    menuDecoration: BoxDecoration(
                      color: Colors.blue,
                      boxShadow: [
                        BoxShadow(
                          color: Colors.black.withOpacity(0.5),
                          blurRadius: 10,
                          spreadRadius: 2,
                        ),
                      ],
                      borderRadius: BorderRadius.circular(10),
                    ),
                    hoverColor: const Color.fromARGB(255, 24, 116, 192),
                    defaultColor: Colors.blue,
                    items: items,
                  ),
                  overlayPosition: JoyOverlayPosition.top,
                  child: const ExampleButton(
                      text: 'Styled Top Position', color: Colors.blue),
                ),
                const SizedBox(width: 20),
                JoyDropDown.list(
                  overlayOffset: const Offset(0, 5),
                  menu: JoyMenu(
                    menuDecoration: BoxDecoration(
                      boxShadow: [
                        BoxShadow(
                          color: Colors.black.withOpacity(0.5),
                          blurRadius: 10,
                          spreadRadius: 2,
                        ),
                      ],
                      borderRadius: BorderRadius.circular(10),
                    ),
                    hoverColor: Color.fromARGB(255, 0, 116, 104),
                    defaultColor: Colors.teal,
                    items: items,
                  ),
                  overlayPosition: JoyOverlayPosition.bottom,
                  child: const ExampleButton(
                      text: 'Styled Bottom Position', color: Colors.teal),
                ),
              ],
            ),
            const SizedBox(height: 60),
            const Text(
              'or try adding anything xD',
              style: TextStyle(fontSize: 40),
            ),
            const SizedBox(height: 20),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                JoyDropDown(
                  overlayHeight: 200,
                  overlayWidth: 200,
                  overlayOffset: const Offset(0, 5),
                  overlayPosition: JoyOverlayPosition.bottom,
                  overlayWidget: Container(
                    clipBehavior: Clip.hardEdge,
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(10),
                    ),
                    child: Image.network(
                        'https://c.tenor.com/2roX3uxz_68AAAAM/cat-space.gif',
                        fit: BoxFit.cover,
                        width: 100,
                        height: 100),
                  ),
                  child: const ExampleButton(
                      text: 'Like this', color: Color.fromARGB(255, 4, 52, 92)),
                ),
                const SizedBox(width: 20),
                JoyDropDown(
                  overlayHeight: 200,
                  overlayWidth: 200,
                  overlayOffset: const Offset(0, 5),
                  overlayPosition: JoyOverlayPosition.bottom,
                  overlayWidget: Container(
                    clipBehavior: Clip.hardEdge,
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(10),
                    ),
                    child: Image.asset(
                      'assets/cavalo-meme.gif',
                      fit: BoxFit.cover,
                      width: 200,
                      height: 200,
                    ),
                  ),
                  child: const ExampleButton(
                    text: 'or this',
                    color: Color.fromARGB(255, 180, 127, 108),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
